tests/std/tests/GH_004597_self_swap/test.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <utility>
+
+struct swap_counter {
+    unsigned int* pcnt_ = nullptr;
+
+    friend _CONSTEXPR20 void swap(swap_counter& lhs, swap_counter& rhs) noexcept {
+        std::swap(lhs.pcnt_, rhs.pcnt_);
+        if (lhs.pcnt_ != nullptr) {
+            ++(*lhs.pcnt_);
+        }
+        if (rhs.pcnt_ != nullptr) {
+            ++(*rhs.pcnt_);
+        }
+    }
+
+    _CONSTEXPR20 bool operator==(unsigned int x) const {
+        return *pcnt_ == x;
+    }
+};
+
+
+_CONSTEXPR20 bool test_gh_4595() {
+    auto res1 = [] {
+        unsigned int cnt{};
+        std::pair<swap_counter, int> pr{swap_counter{&cnt}, 0};
+        pr.swap(pr);
+        return (cnt == 2u) && (pr.first == 2u) && (pr.second == 0);

```

Comment:
The parentheses here are unnecessary - we consider equality to have well-known higher precedence than logical operators. Occurs below.